File: Docs/examples/python/trading_client.py
"""
GRID Trading API Client (Python)

Complete client for interacting with the GRID Trading API
"""

import json
import requests
import time
from typing import Optional, Dict, List, Any
from urllib.parse import urlencode

from auth import SignatureAuth


class GridTradingClient:
    """
    Complete client for GRID Trading API
    """
    
    def __init__(
        self,
        private_key: str,
        public_key: str,
        base_url: str = 'https://trading.api.thegrid.ai/v1'
    ):
        """
        Initialize the trading client
        
        Args:
            private_key: Base64-encoded Ed25519 private key
            public_key: Base64-encoded Ed25519 public key
            base_url: API base URL
        """
        self.auth = SignatureAuth(private_key, public_key)
        self.base_url = base_url
        self.session = requests.Session()
        self.session.headers.update({'Content-Type': 'application/json'})
    
    def _request(
        self,
        method: str,
        path: str,
        data: Optional[Dict] = None,
        params: Optional[Dict] = None
    ) -> Any:
        """
        Make an authenticated API request
        
        Args:
            method: HTTP method
            path: Request path
            data: Request body (for POST/PUT)
            params: Query parameters (for GET)
        
        Returns:
            Response data
        
        Raises:
            requests.HTTPError: If request fails
        """
        body = json.dumps(data) if data else ''
        headers = self.auth.get_headers(method, path, body)
        
        url = f"{self.base_url}{path}"
        
        response = self.session.request(
            method=method,
            url=url,
            headers=headers,
            json=data,
            params=params
        )
        
        response.raise_for_status()
        return response.json()
    
    # ==================== Markets ====================
    
    def get_markets(self) -> List[Dict]:
        """
        List all available markets
        
        Returns:
            List of market objects
        """
        response = self._request('GET', '/trading/markets')
        return response['data']
    
    def get_market(self, market_id: str) -> Dict:
        """
        Get details for a specific market
        
        Args:
            market_id: Market identifier
        
        Returns:
            Market details
        """
        response = self._request('GET', f'/trading/markets/{market_id}')
        return response['data']
    
    def get_ticker(self, market_id: str) -> Dict:
        """
        Get ticker data for a market
        
        Args:
            market_id: Market identifier
        
        Returns:
            Ticker data
        """
        response = self._request('GET', f'/trading/markets/{market_id}/ticker')
        return response['data']
    
    def get_order_book(self, market_id: str, depth: int = 20) -> Dict:
        """
        Get order book for a market
        
        Args:
            market_id: Market identifier
            depth: Number of price levels (default: 20, max: 100)
        
        Returns:
            Order book with bids and asks
        """
        response = self._request(
            'GET',
            f'/trading/markets/{market_id}/orderbook',
            params={'depth': depth}
        )
        return response['data']
    
    def get_market_trades(self, market_id: str, limit: int = 50) -> List[Dict]:
        """
        Get recent trades for a market
        
        Args:
            market_id: Market identifier
            limit: Number of trades (default: 50, max: 100)
        
        Returns:
            List of trade objects
        """
        response = self._request(
            'GET',
            f'/trading/markets/{market_id}/trades',
            params={'limit': limit}
        )
        return response['data']
    
    # ==================== Orders ====================
    
    def place_order(
        self,
        market_id: str,
        side: str,
        order_type: str,
        quantity: str,
        price: Optional[str] = None,
        time_in_force: str = 'gtc',
        client_order_id: Optional[str] = None
    ) -> Dict:
        """
        Place a new order
        
        Args:
            market_id: Market identifier
            side: 'buy' or 'sell'
            order_type: 'limit' or 'market'
            quantity: Order quantity as string
            price: Limit price (required for limit orders)
            time_in_force: 'gtc', 'ioc', 'fok', or 'day'
            client_order_id: Optional client-specified ID
        
        Returns:
            Created order object
        """
        order_data = {
            'market_id': market_id,
            'side': side,
            'type': order_type,
            'quantity': quantity,
            'time_in_force': time_in_force
        }
        
        if price:
            order_data['price'] = price
        
        if client_order_id:
            order_data['client_order_id'] = client_order_id
        
        response = self._request('POST', '/trading/orders', data=order_data)
        return response['data']
    
    def list_orders(
        self,
        status: Optional[str] = None,
        market_id: Optional[str] = None,
        limit: int = 50,
        offset: int = 0
    ) -> List[Dict]:
        """
        List orders with optional filters
        
        Args:
            status: 'open', 'filled', 'cancelled', or 'all'
            market_id: Filter by market
            limit: Results per page (default: 50)
            offset: Pagination offset
        
        Returns:
            List of order objects
        """
        params = {'limit': limit, 'offset': offset}
        if status:
            params['status'] = status
        if market_id:
            params['market_id'] = market_id
        
        response = self._request('GET', '/trading/orders', params=params)
        return response['data']
    
    def get_order(self, order_id: str) -> Dict:
        """
        Get details for a specific order
        
        Args:
            order_id: Order identifier
        
        Returns:
            Order details
        """
        response = self._request('GET', f'/trading/orders/{order_id}')
        return response['data']
    
    def cancel_order(self, order_id: str) -> Dict:
        """
        Cancel an existing order
        
        Args:
            order_id: Order identifier
        
        Returns:
            Cancellation result
        """
        response = self._request('DELETE', f'/trading/orders/{order_id}')
        return response['data']
    
    def cancel_all_orders(self, market_id: Optional[str] = None) -> Dict:
        """
        Cancel all open orders (optionally filter by market)
        
        Args:
            market_id: Optional market filter
        
        Returns:
            Cancellation results
        """
        orders = self.list_orders(status='open', market_id=market_id)
        
        results = []
        for order in orders:
            try:
                result = self.cancel_order(order['order_id'])
                results.append({'success': True, 'order_id': order['order_id']})
            except Exception as e:
                results.append({
                    'success': False,
                    'order_id': order['order_id'],
                    'error': str(e)
                })
        
        return {
            'total': len(orders),
            'successful': len([r for r in results if r['success']]),
            'failed': len([r for r in results if not r['success']]),
            'results': results
        }
    
    # ==================== Trades ====================
    
    def get_trades(
        self,
        market_id: Optional[str] = None,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        limit: int = 50
    ) -> List[Dict]:
        """
        Get user trade history
        
        Args:
            market_id: Filter by market
            start_date: ISO 8601 start date
            end_date: ISO 8601 end date
            limit: Results per page
        
        Returns:
            List of trade objects
        """
        params = {'limit': limit}
        if market_id:
            params['market_id'] = market_id
        if start_date:
            params['start_date'] = start_date
        if end_date:
            params['end_date'] = end_date
        
        response = self._request('GET', '/trading/trades', params=params)
        return response['data']
    
    def get_trade(self, trade_id: str) -> Dict:
        """
        Get details for a specific trade
        
        Args:
            trade_id: Trade identifier
        
        Returns:
            Trade details
        """
        response = self._request('GET', f'/trading/trades/{trade_id}')
        return response['data']
    
    # ==================== Accounts ====================
    
    def get_trading_accounts(self) -> List[Dict]:
        """
        Get trading account balances
        
        Returns:
            List of account balance objects
        """
        response = self._request('GET', '/trading/trading-accounts')
        return response['data']
    
    def get_trading_account(self, account_id: str) -> Dict:
        """
        Get specific trading account
        
        Args:
            account_id: Account identifier
        
        Returns:
            Account details
        """
        response = self._request('GET', f'/trading/trading-accounts/{account_id}')
        return response['data']
    
    def get_currency_trading_accounts(self) -> List[Dict]:
        """
        Get currency trading accounts
        
        Returns:
            List of currency trading account objects
        """
        response = self._request('GET', '/trading/currency-trading-accounts')
        return response['data']
    
    # ==================== Positions ====================
    
    def get_positions(self, status: Optional[str] = None) -> List[Dict]:
        """
        Get user positions
        
        Args:
            status: Filter by status ('open' or 'closed')
        
        Returns:
            List of position objects
        """
        params = {}
        if status:
            params['filters[0][field]'] = 'status'
            params['filters[0][value]'] = status
        
        response = self._request('GET', '/positions', params=params)
        return response['data']
    
    # ==================== Price History ====================
    
    def get_price_history(
        self,
        market_id: str,
        resolution: str,
        from_timestamp: int,
        to_timestamp: int
    ) -> List[Dict]:
        """
        Get price history (OHLCV candles) for a market
        
        Args:
            market_id: Market identifier
            resolution: Candle resolution (1m, 5m, 15m, 30m, 1h, 4h, 1d, 1w, 1M)
            from_timestamp: Start time (Unix timestamp)
            to_timestamp: End time (Unix timestamp)
        
        Returns:
            List of price candle objects
        """
        params = {
            'filters[0][field]': 'market_id',
            'filters[0][value]': market_id,
            'filters[1][field]': 'resolution',
            'filters[1][value]': resolution,
            'filters[2][field]': 'from',
            'filters[2][value]': str(from_timestamp),
            'filters[3][field]': 'to',
            'filters[3][value]': str(to_timestamp),
            'order_by[]': 'period_start',
            'order_directions[]': 'asc'
        }
        
        response = self._request('GET', '/price_histories', params=params)
        return response['data']
    
    # ==================== Instruments ====================
    
    def list_instruments(self) -> List[Dict]:
        """
        List all instruments (public endpoint - no auth required)
        
        Returns:
            List of instrument objects
        """
        response = requests.get(f'{self.base_url}/instruments')
        response.raise_for_status()
        return response.json()['data']
    
    def get_instrument(self, instrument_id: str) -> Dict:
        """
        Get instrument details by ID
        
        Args:
            instrument_id: Instrument identifier
        
        Returns:
            Instrument details
        """
        response = requests.get(f'{self.base_url}/instruments/{instrument_id}')
        response.raise_for_status()
        return response.json()['data']
    
    def get_instrument_by_symbol(self, symbol: str) -> Dict:
        """
        Get instrument by symbol
        
        Args:
            symbol: Instrument symbol
        
        Returns:
            Instrument details
        """
        response = requests.get(f'{self.base_url}/instruments/by-symbol/{symbol}')
        response.raise_for_status()
        return response.json()['data']
    
    # ==================== Market Stats ====================
    
    def get_market_stats(self, market_id: str) -> Dict:
        """
        Get 24-hour market statistics
        
        Args:
            market_id: Market identifier
        
        Returns:
            Market statistics
        """
        response = self._request('GET', f'/markets/{market_id}/stats')
        return response['data']
    
    # ==================== Transfers ====================
    
    def transfer_to_consumption(self, instrument_id: str, quantity: int) -> Dict:
        """
        Transfer from trading to consumption account
        
        Args:
            instrument_id: Instrument to transfer
            quantity: Amount to transfer
        
        Returns:
            Transfer result with transfer_id
        """
        response = self._request(
            'POST',
            '/transfers/trading-to-consumption',
            data={'instrument_id': instrument_id, 'quantity': quantity}
        )
        return response
    
    def transfer_to_trading(self, instrument_id: str, quantity: int) -> Dict:
        """
        Transfer from consumption to trading account
        
        Args:
            instrument_id: Instrument to transfer
            quantity: Amount to transfer
        
        Returns:
            Transfer result with transfer_id
        """
        response = self._request(
            'POST',
            '/transfers/consumption-to-trading',
            data={'instrument_id': instrument_id, 'quantity': quantity}
        )
        return response
    
    def get_transfer_history(
        self,
        market_id: Optional[str] = None,
        limit: int = 50
    ) -> List[Dict]:
        """
        Get transfer history
        
        Args:
            market_id: Filter by market (optional)
            limit: Results per page
        
        Returns:
            List of transfer records
        """
        params = {'page_size': limit}
        if market_id:
            params['market_id'] = market_id
        
        response = self._request('GET', '/transfers/histories', params=params)
        return response['data']
    
    # ==================== Consumption ====================
    
    def get_consumption_instruments(self, api_key_id: Optional[str] = None) -> List[Dict]:
        """
        Get consumption instruments (balance and usage)
        
        Args:
            api_key_id: Filter by API key (optional)
        
        Returns:
            List of consumption instrument objects
        """
        params = {}
        if api_key_id:
            params['filters[0][field]'] = 'api_key_id'
            params['filters[0][value]'] = api_key_id
        
        response = self._request('GET', '/consumption/instruments', params=params)
        return response['data']
    
    # ==================== Public Data ====================
    
    def get_public_trades(self, market_id: str, limit: int = 50) -> List[Dict]:
        """
        Get public trades (no authentication required)
        
        Args:
            market_id: Filter by market
            limit: Results per page
        
        Returns:
            List of public trade objects
        """
        params = {
            'filters[0][field]': 'market_id',
            'filters[0][value]': market_id,
            'order_by[]': 'execution_timestamp',
            'order_directions[]': 'desc',
            'page_size': limit
        }
        
        response = requests.get(
            f'{self.base_url}/public_trades',
            params=params
        )
        response.raise_for_status()
        return response.json()['data']


# Example usage
if __name__ == '__main__':
    # Load keys from files
    with open('./ed25519.key', 'r') as f:
        private_key = f.read().strip()
    
    with open('./ed25519_pub.der', 'r') as f:
        public_key = f.read().strip()
    
    # Initialize client
    client = GridTradingClient(private_key, public_key)
    
    try:
        # Get markets
        markets = client.get_markets()
        print(f"Available markets: {len(markets)}")
        
        # Get balances
        balances = client.get_trading_accounts()
        print(f"Balances: {len(balances)} accounts")
        
        # Get positions
        positions = client.get_positions(status='open')
        print(f"Open positions: {len(positions)}")
        
        # List instruments
        instruments = client.list_instruments()
        ai_instruments = [i for i in instruments if i['instrument_type'] == 'ai_commodity']
        print(f"Available instruments: {len(instruments)} ({len(ai_instruments)} AI commodities)")
        
        # Get consumption balance
        consumption = client.get_consumption_instruments()
        if consumption:
            print(f"Consumption instruments: {len(consumption)}")
            for inst in consumption:
                usage_pct = (inst['tokens_used'] / inst['tokens_purchased']) * 100
                print(f"  {inst['instrument_name']}: {usage_pct:.1f}% used")
        
        # Get price history (last 24 hours, hourly candles)
        if markets:
            market_id = markets[0]['market_id']
            now = int(time.time())
            yesterday = now - (24 * 60 * 60)
            
            candles = client.get_price_history(market_id, '1h', yesterday, now)
            print(f"Price history: {len(candles)} hourly candles")
        
        # Place an order (commented out for safety)
        # order = client.place_order(
        #     market_id='market_b310e860-97cd-45eb-bdc3-5be0b79295d0',
        #     side='buy',
        #     order_type='limit',
        #     quantity='0.01',
        #     price='45000.00'
        # )
        # print(f"Order placed: {order}")
        
    except requests.HTTPError as e:
        print(f"API Error: {e.response.status_code} - {e.response.text}")
    except Exception as e:
        print(f"Error: {e}")

File: Docs/examples/javascript/trading-client.js
/**
 * GRID Trading API Client (JavaScript/TypeScript)
 * 
 * Complete client for interacting with the GRID Trading API
 */

const axios = require('axios');
const { SignatureAuth } = require('./auth');

class GridTradingClient {
  /**
   * Initialize the trading client
   * 
   * @param {string} privateKey - Base64-encoded Ed25519 private key
   * @param {string} publicKey - Base64-encoded Ed25519 public key
   * @param {Object} options - Optional configuration
   * @param {string} options.baseUrl - API base URL
   */
  constructor(privateKey, publicKey, options = {}) {
    this.auth = new SignatureAuth(privateKey, publicKey);
    this.baseUrl = options.baseUrl || 'https://trading.api.thegrid.ai/v1';
    
    this.client = axios.create({
      baseURL: this.baseUrl,
      headers: {
        'Content-Type': 'application/json'
      }
    });
  }

  /**
   * Make an authenticated API request
   * 
   * @private
   */
  async _request(method, path, data = null) {
    const body = data ? JSON.stringify(data) : '';
    const headers = this.auth.getHeaders(method, path, body);
    
    const config = {
      method,
      url: path,
      headers,
      ...(data && { data })
    };
    
    try {
      const response = await this.client.request(config);
      return response.data;
    } catch (error) {
      if (error.response) {
        throw new Error(`API Error: ${error.response.status} - ${JSON.stringify(error.response.data)}`);
      }
      throw error;
    }
  }

  // ==================== Markets ====================

  /**
   * List all available markets
   * 
   * @returns {Promise<Array>} Array of market objects
   */
  async getMarkets() {
    const response = await this._request('GET', '/trading/markets');
    return response.data;
  }

  /**
   * Get details for a specific market
   * 
   * @param {string} marketId - Market identifier
   * @returns {Promise<Object>} Market details
   */
  async getMarket(marketId) {
    const response = await this._request('GET', `/trading/markets/${marketId}`);
    return response.data;
  }

  /**
   * Get ticker data for a market
   * 
   * @param {string} marketId - Market identifier
   * @returns {Promise<Object>} Ticker data
   */
  async getTicker(marketId) {
    const response = await this._request('GET', `/trading/markets/${marketId}/ticker`);
    return response.data;
  }

  /**
   * Get order book for a market
   * 
   * @param {string} marketId - Market identifier
   * @param {number} depth - Number of price levels (default: 20, max: 100)
   * @returns {Promise<Object>} Order book with bids and asks
   */
  async getOrderBook(marketId, depth = 20) {
    const response = await this._request('GET', `/trading/markets/${marketId}/orderbook?depth=${depth}`);
    return response.data;
  }

  /**
   * Get recent trades for a market
   * 
   * @param {string} marketId - Market identifier
   * @param {number} limit - Number of trades (default: 50, max: 100)
   * @returns {Promise<Array>} Array of trade objects
   */
  async getMarketTrades(marketId, limit = 50) {
    const response = await this._request('GET', `/trading/markets/${marketId}/trades?limit=${limit}`);
    return response.data;
  }

  // ==================== Orders ====================

  /**
   * Place a new order
   * 
   * @param {Object} order - Order parameters
   * @param {string} order.market_id - Market identifier
   * @param {string} order.side - 'buy' or 'sell'
   * @param {string} order.type - 'limit' or 'market'
   * @param {string} order.quantity - Order quantity as string
   * @param {string} order.price - Limit price (required for limit orders)
   * @param {string} order.time_in_force - 'gtc', 'ioc', 'fok', or 'day'
   * @param {string} order.client_order_id - Optional client-specified ID
   * @returns {Promise<Object>} Created order object
   */
  async placeOrder(order) {
    const response = await this._request('POST', '/trading/orders', order);
    return response.data;
  }

  /**
   * List orders with optional filters
   * 
   * @param {Object} filters - Filter parameters
   * @param {string} filters.status - 'open', 'filled', 'cancelled', or 'all'
   * @param {string} filters.market_id - Filter by market
   * @param {number} filters.limit - Results per page (default: 50)
   * @param {number} filters.offset - Pagination offset
   * @returns {Promise<Array>} Array of order objects
   */
  async listOrders(filters = {}) {
    const queryString = new URLSearchParams(filters).toString();
    const path = `/trading/orders${queryString ? '?' + queryString : ''}`;
    const response = await this._request('GET', path);
    return response.data;
  }

  /**
   * Get details for a specific order
   * 
   * @param {string} orderId - Order identifier
   * @returns {Promise<Object>} Order details
   */
  async getOrder(orderId) {
    const response = await this._request('GET', `/trading/orders/${orderId}`);
    return response.data;
  }

  /**
   * Cancel an existing order
   * 
   * @param {string} orderId - Order identifier
   * @returns {Promise<Object>} Cancellation result
   */
  async cancelOrder(orderId) {
    const response = await this._request('DELETE', `/trading/orders/${orderId}`);
    return response.data;
  }

  /**
   * Cancel all open orders (optionally filter by market)
   * 
   * @param {string} marketId - Optional market filter
   * @returns {Promise<Object>} Cancellation results
   */
  async cancelAllOrders(marketId = null) {
    const filters = { status: 'open' };
    if (marketId) filters.market_id = marketId;
    
    const orders = await this.listOrders(filters);
    const cancelPromises = orders.map(order => 
      this.cancelOrder(order.order_id).catch(err => ({
        error: true,
        orderId: order.order_id,
        message: err.message
      }))
    );
    
    const results = await Promise.all(cancelPromises);
    return {
      total: orders.length,
      successful: results.filter(r => !r.error).length,
      failed: results.filter(r => r.error).length,
      results
    };
  }

  // ==================== Trades ====================

  /**
   * Get user trade history
   * 
   * @param {Object} filters - Filter parameters
   * @param {string} filters.market_id - Filter by market
   * @param {string} filters.start_date - ISO 8601 start date
   * @param {string} filters.end_date - ISO 8601 end date
   * @param {number} filters.limit - Results per page
   * @returns {Promise<Array>} Array of trade objects
   */
  async getTrades(filters = {}) {
    const queryString = new URLSearchParams(filters).toString();
    const path = `/trading/trades${queryString ? '?' + queryString : ''}`;
    const response = await this._request('GET', path);
    return response.data;
  }

  /**
   * Get details for a specific trade
   * 
   * @param {string} tradeId - Trade identifier
   * @returns {Promise<Object>} Trade details
   */
  async getTrade(tradeId) {
    const response = await this._request('GET', `/trading/trades/${tradeId}`);
    return response.data;
  }

  // ==================== Accounts ====================

  /**
   * Get trading account balances
   * 
   * @returns {Promise<Array>} Array of account balance objects
   */
  async getTradingAccounts() {
    const response = await this._request('GET', '/trading/trading-accounts');
    return response.data;
  }

  /**
   * Get specific trading account
   * 
   * @param {string} accountId - Account identifier
   * @returns {Promise<Object>} Account details
   */
  async getTradingAccount(accountId) {
    const response = await this._request('GET', `/trading/trading-accounts/${accountId}`);
    return response.data;
  }

  /**
   * Get currency trading accounts
   * 
   * @returns {Promise<Array>} Currency trading account objects
   */
  async getCurrencyTradingAccounts() {
    const response = await this._request('GET', '/trading/currency-trading-accounts');
    return response.data;
  }

  // ==================== Positions ====================

  /**
   * Get user positions
   * 
   * @param {Object} filters - Filter parameters
   * @param {string} filters.status - Filter by status ('open' or 'closed')
   * @returns {Promise<Array>} Array of position objects
   */
  async getPositions(filters = {}) {
    const params = {};
    let filterIdx = 0;
    
    if (filters.status) {
      params[`filters[${filterIdx}][field]`] = 'status';
      params[`filters[${filterIdx}][value]`] = filters.status;
    }
    
    const queryString = new URLSearchParams(params).toString();
    const path = `/positions${queryString ? '?' + queryString : ''}`;
    const response = await this._request('GET', path);
    return response.data;
  }

  // ==================== Price History ====================

  /**
   * Get price history (OHLCV candles) for a market
   * 
   * @param {string} marketId - Market identifier
   * @param {string} resolution - Candle resolution (1m, 5m, 15m, 30m, 1h, 4h, 1d, 1w, 1M)
   * @param {number} fromTimestamp - Start time (Unix timestamp)
   * @param {number} toTimestamp - End time (Unix timestamp)
   * @returns {Promise<Array>} Array of price candles
   */
  async getPriceHistory(marketId, resolution, fromTimestamp, toTimestamp) {
    const params = {
      'filters[0][field]': 'market_id',
      'filters[0][value]': marketId,
      'filters[1][field]': 'resolution',
      'filters[1][value]': resolution,
      'filters[2][field]': 'from',
      'filters[2][value]': fromTimestamp.toString(),
      'filters[3][field]': 'to',
      'filters[3][value]': toTimestamp.toString(),
      'order_by[]': 'period_start',
      'order_directions[]': 'asc'
    };
    
    const queryString = new URLSearchParams(params).toString();
    const path = `/price_histories?${queryString}`;
    const response = await this._request('GET', path);
    return response.data;
  }

  // ==================== Instruments ====================

  /**
   * List all instruments (public endpoint - no auth required)
   * 
   * @returns {Promise<Array>} Array of instrument objects
   */
  async listInstruments() {
    try {
      const response = await axios.get(`${this.baseUrl}/instruments`);
      return response.data.data;
    } catch (error) {
      throw new Error(`Failed to list instruments: ${error.message}`);
    }
  }

  /**
   * Get instrument details by ID
   * 
   * @param {string} instrumentId - Instrument identifier
   * @returns {Promise<Object>} Instrument details
   */
  async getInstrument(instrumentId) {
    try {
      const response = await axios.get(`${this.baseUrl}/instruments/${instrumentId}`);
      return response.data.data;
    } catch (error) {
      throw new Error(`Failed to get instrument: ${error.message}`);
    }
  }

  /**
   * Get instrument by symbol
   * 
   * @param {string} symbol - Instrument symbol
   * @returns {Promise<Object>} Instrument details
   */
  async getInstrumentBySymbol(symbol) {
    try {
      const response = await axios.get(`${this.baseUrl}/instruments/by-symbol/${symbol}`);
      return response.data.data;
    } catch (error) {
      throw new Error(`Failed to get instrument by symbol: ${error.message}`);
    }
  }

  // ==================== Market Stats ====================

  /**
   * Get 24-hour market statistics
   * 
   * @param {string} marketId - Market identifier
   * @returns {Promise<Object>} Market statistics
   */
  async getMarketStats(marketId) {
    const response = await this._request('GET', `/markets/${marketId}/stats`);
    return response.data;
  }

  // ==================== Transfers ====================

  /**
   * Transfer from trading to consumption account
   * 
   * @param {string} instrumentId - Instrument to transfer
   * @param {number} quantity - Amount to transfer
   * @returns {Promise<Object>} Transfer result
   */
  async transferToConsumption(instrumentId, quantity) {
    const response = await this._request('POST', '/transfers/trading-to-consumption', {
      instrument_id: instrumentId,
      quantity: quantity
    });
    return response;
  }

  /**
   * Transfer from consumption to trading account
   * 
   * @param {string} instrumentId - Instrument to transfer
   * @param {number} quantity - Amount to transfer
   * @returns {Promise<Object>} Transfer result
   */
  async transferToTrading(instrumentId, quantity) {
    const response = await this._request('POST', '/transfers/consumption-to-trading', {
      instrument_id: instrumentId,
      quantity: quantity
    });
    return response;
  }

  /**
   * Get transfer history
   * 
   * @param {string} marketId - Filter by market (optional)
   * @param {number} limit - Results per page
   * @returns {Promise<Array>} Transfer history
   */
  async getTransferHistory(marketId = null, limit = 50) {
    const params = { page_size: limit };
    if (marketId) params.market_id = marketId;
    
    const queryString = new URLSearchParams(params).toString();
    const path = `/transfers/histories?${queryString}`;
    const response = await this._request('GET', path);
    return response.data;
  }

  // ==================== Consumption ====================

  /**
   * Get consumption instruments (balance and usage)
   * 
   * @param {string} apiKeyId - Filter by API key (optional)
   * @returns {Promise<Array>} Consumption instruments
   */
  async getConsumptionInstruments(apiKeyId = null) {
    const params = {};
    if (apiKeyId) {
      params['filters[0][field]'] = 'api_key_id';
      params['filters[0][value]'] = apiKeyId;
    }
    
    const queryString = new URLSearchParams(params).toString();
    const path = `/consumption/instruments${queryString ? '?' + queryString : ''}`;
    const response = await this._request('GET', path);
    return response.data;
  }

  // ==================== Public Data ====================

  /**
   * Get public trades (no authentication required)
   * 
   * @param {string} marketId - Filter by market
   * @param {number} limit - Results per page
   * @returns {Promise<Array>} Public trades
   */
  async getPublicTrades(marketId, limit = 50) {
    const params = {
      'filters[0][field]': 'market_id',
      'filters[0][value]': marketId,
      'order_by[]': 'execution_timestamp',
      'order_directions[]': 'desc',
      'page_size': limit
    };
    
    try {
      const response = await axios.get(
        `${this.baseUrl}/public_trades`,
        { params }
      );
      return response.data.data;
    } catch (error) {
      throw new Error(`Failed to get public trades: ${error.message}`);
    }
  }
}

module.exports = { GridTradingClient };

// Example usage:
if (require.main === module) {
  const fs = require('fs');
  
  // Load keys from files
  const privateKey = fs.readFileSync('./ed25519.key', 'utf8').trim();
  const publicKey = fs.readFileSync('./ed25519_pub.der', 'utf8').trim();
  
  const client = new GridTradingClient(privateKey, publicKey);
  
  (async () => {
    try {
      // Get markets
      const markets = await client.getMarkets();
      console.log('Available markets:', markets.length);
      
      // Get balances
      const balances = await client.getTradingAccounts();
      console.log('Balances:', balances);
      
      // Place an order
      // const order = await client.placeOrder({
      //   market_id: 'market_b310e860-97cd-45eb-bdc3-5be0b79295d0',
      //   side: 'buy',
      //   type: 'limit',
      //   quantity: '0.01',
      //   price: '45000.00',
      //   time_in_force: 'gtc'
      // });
      // console.log('Order placed:', order);
      
    } catch (error) {
      console.error('Error:', error.message);
    }
  })();
}

